fair/src/autogluon/fair/learners/fair.py

```diff
@@ -0,0 +1,418 @@
+        groups may be:
+            1. The name of a pandas column. 
+            2. a sequence of binary masks.
+            3. a vector where each unique value indicates membership of a different group
+            4. a function that returns per class confidence scores.
+               in this case, the maximum valued class is taken as a datapoints label
+            5. The value false (used when we don't require groups, for example,
+               if we are optimizing F1) #Not implemented
+               
+        The classifier is evaluated/altered on validation_data. Where this is None and 
+        the classifier is an autogluon classifier, default back to their validation set.
+        Infered groups is a classifier who's responses are treated as per datapoint weights
+        used to scale the thresholds infered by maximize. """
+        self.classifier = classifier
+        self.groups = groups
+        
+        #TODO: handle absense of validation data
+        #if validation_data is None:
+        #    if isinstance(classifier, ag.autogluon): #todo what is the type of an ag.classifier?
+        #        validation_data=classifier.held_back_data #ToDo what is the right name?
+        #    else:
+        #        warning('Fair was constructed with no validation data and no autogluon classifier. validation_data is not initialized. Aborting construction')
+        #        return
+        self.validation_data = validation_data
+
+
+        # We use _internal_groups as a standardized argument that is always safe to pass
+        # to functions expecting either a vector or list of masks
+        if callable(groups):
+            self._internal_groups = groups(validation_data).argmax(1)
+        elif isinstance(groups, str):
+            self._internal_groups = np.asarray(validation_data[groups])
+        else:
+            self._internal_groups = self.groups
+        #ToDo add dim check here.
+
+        self.infered_groups=infered_groups
+        if infered_groups:
+            if callable(infered_groups):
+                self._val_thresholds=infered_groups(validation_data)
+            else:#assume autogluon predictor
+                self._val_thresholds=np.asarray(infered_groups.predict_proba(validation_data))
+                #might want a final case of passed array
+                #assert(infered_groups.ndim==2)
+                #self._val_thresholds=infered_groups
+        else:
+            from sklearn.preprocessing import OneHotEncoder
+            #Use OneHot and store encoder so it will work on new data
+            self.group_encoder=OneHotEncoder(handle_unknown='ignore')
+            self.group_encoder.fit(self._internal_groups.reshape(-1, 1))
+            self._val_thresholds=self.group_encoder.transform(self._internal_groups.reshape(-1, 1)).toarray()
+        
+        if callable(classifier):
+            self.proba=classifier(validation_data)
+            #ToDo: scikit-learn
```

Comment:
This opens up a massive can of worms. Please don't do this unless we are willing to go all the way with it.